lib/dynamic_config/datastore_cache.rb

```diff
@@ -1,112 +1,104 @@
+require 'cdo/shared_cache'
 require 'honeybadger/ruby'
 
-# A caching layer that sits in front of a datastore that
-# implements get and set
-
+# A multilayered caching interface intended for use by Cdo::DynamicConfig. Ties
+# together three different components:
+#
+#   1. The long-lived datastore, which in production is DynamoDB. This is the
+#      ultimate source of truth for the current state of our configuration
+#      settings, and will only be read from on initialization and when a
+#      configuration value is updated.
+#
+#   2. The shared cache, which in production is ElastiCache. This is a middle
+#      ground shared between the frontend processes which will be kept up to
+#      date with the content in the datastore and from which processes will
+#      fetch updates on a configurable schedule.
+#
+#   3. The local in-memory cache, which in all environments is a simple hash.
+#      Each frontend processes will reference its local cache to serve the
+#      vast majority of user requests, and will occasionally update its
+#      contents from the shared cache.
```

Comment:
I /really/ appreciate this comment, IMO the PR is a "merge yes please!" based on this comment alone haha.

Reply:
For future code archeologists, please do read the "previous attempt PR" @Hamms mentions as well: https://github.com/code-dot-org/code-dot-org/pull/53833

This gives the full motivating context for this change (TL;DR: we depended on unsafe thread behavior in ruby <2, which actually manifests as problems in ruby 3)